### PR TITLE
Vrtcal adapter: Added Native Support

### PR DESCRIPTION
**Pull request by vrtcal-dev**

### CODE REVIEW

File: static/bidder-info/vrtcal.yaml [hunk 1]
```diff
@@ -7,13 +7,14 @@ capabilities:
     mediaTypes:
       - banner
       - video
+      - native
   site:
     mediaTypes:
       - banner
       - video
+      - native
```
Comment (gargcreation1992) [inline static/bidder-info/vrtcal.yaml:15]: Requesting you to please add the json test case for this change.

File: adapters/vrtcal/vrtcal.go [hunk 2]
```diff
@@ -101,10 +101,17 @@ func getReturnTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType
...
 
+
 			if imp.Video != nil {
 				return openrtb_ext.BidTypeVideo, nil
 			}
 
+
+			if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
+			}
```
Comment (gargcreation1992) [inline adapters/vrtcal/vrtcal.go:112]: Warning - This is an anti-pattern. This code assumes that for multi-format bid request, mediatype will be set to type by default. I would recommend to set the mediaType based on bidder response. Your bidder should be able to set `openrtb2.Bid.MType` field and use that field here to get mediaTypeForImp. [Here is an example](https://github.com/prebid/prebid-server/blob/f275f3287f644293d102c07be4306a6b81cf2642/adapters/aidem/aidem.go#L89) on how this is done. Even if your bidder does not support multi-format request, I would highly encourage you to make this change.

File: adapters/vrtcal/vrtcal.go [hunk 3]
```diff
@@ -101,10 +101,17 @@ func getReturnTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType
 				return openrtb_ext.BidTypeBanner, nil
 			}
 
+
```
Comment (gargcreation1992) [inline adapters/vrtcal/vrtcal.go:104]: nitpick - Please remove this blank line.

File: adapters/vrtcal/vrtcal.go [hunk 4]
```diff
@@ -101,10 +101,17 @@ func getReturnTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType
...
 			if imp.Video != nil {
 				return openrtb_ext.BidTypeVideo, nil
 			}
 
+
```
Comment (gargcreation1992) [inline adapters/vrtcal/vrtcal.go:109]: nit - there is already a blank line above line 109. Please remove this blank line.

File: adapters/vrtcal/vrtcal.go [hunk 5]
```diff
@@ -101,10 +101,17 @@ func getReturnTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType
...
+			if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
+			}
+
+
```
Comment (gargcreation1992) [inline adapters/vrtcal/vrtcal.go:114]: nitpick - Please remove this blank line.

**Review by onkarvhanumante — DISMISSED** [review]
g2g after addressing @gargcreation1992  comments

**Comment by gargcreation1992** [comment]
@vrtcal-dev Requesting you to please address the open comments.

File: adapters/vrtcal/vrtcal.go (hunk 2 above)
Comment (vrtcal-dev) [reply adapters/vrtcal/vrtcal.go]: Thank you for the feedback...We've updated to support the 2.6 mtype parameter when present....We've also included code to still support the pre 2.6 non passed mtype (being it wasnt in the oRTB standard until 2.6).

File: static/bidder-info/vrtcal.yaml (hunk 1 above)
Comment (vrtcal-dev) [reply static/bidder-info/vrtcal.yaml:15]: Native test cases added

File: adapters/vrtcal/vrtcal.go [hunk 6]
```diff
@@ -94,25 +94,37 @@ func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server co
 	return bidder, nil
 }
 
-func getReturnTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+func getReturnTypeForImp(MType openrtb2.MarkupType, impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
```
Comment (onkarvhanumante) [inline adapters/vrtcal/vrtcal.go:97]: should use camelCase style for var. `MType` -> `mType`

File: adapters/vrtcal/vrtcal.go [hunk 7]
```diff
@@ -94,25 +94,37 @@ func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server co
...
+				}
+				if imp.Video != nil {
+					return openrtb_ext.BidTypeVideo, nil
+				}
+				if imp.Native != nil {
+					return openrtb_ext.BidTypeNative, nil
+				} else {
+					return "", &errortypes.BadServerResponse{
+						Message: fmt.Sprintf("Unsupported return type for ID: (mtype not present)\"%s\"", impID)}
+				}
```
Comment (onkarvhanumante) [inline adapters/vrtcal/vrtcal.go:123]: Error returned from line 109 has same message. Should update either of them

**Review by onkarvhanumante — DISMISSED** [review]

File: adapters/vrtcal/vrtcal.go [hunk 8]
```diff
@@ -94,25 +94,37 @@ func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server co
...
-				return openrtb_ext.BidTypeBanner, nil
-			}
-
-			if imp.Video != nil {
-				return openrtb_ext.BidTypeVideo, nil
-			}
-
-			return "", &errortypes.BadServerResponse{
-				Message: fmt.Sprintf("Unsupported return type for ID: \"%s\"", impID),
+			if mType != 0 {
```
Comment (gargcreation1992) [inline adapters/vrtcal/vrtcal.go:100]: This part of code does not have to be in for loop. This function argument is having `mType` as input that is set on seatBid.Bid[i] which can only hold one impression. I would suggest to remove this for loop.

**Review by gargcreation1992 — COMMENTED** [review]
requesting few changes.

File: adapters/vrtcal/vrtcal.go (hunk 8 above)
Comment (vrtcal-dev) [reply adapters/vrtcal/vrtcal.go]: Cleaned up this function to reflect your suggestions and updated related test cases.

File: adapters/vrtcal/vrtcal.go [hunk 9]
```diff
@@ -94,25 +94,37 @@ func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server co
...
+					return openrtb_ext.BidTypeBanner, nil
+				} else if mType == openrtb2.MarkupVideo {
+					return openrtb_ext.BidTypeVideo, nil
+				} else if mType == openrtb2.MarkupNative {
+					return openrtb_ext.BidTypeNative, nil
+				} else {
+					return "", &errortypes.BadServerResponse{
+						Message: fmt.Sprintf("Unsupported return type for ID (mtype present): \"%s\"", impID)}
+				}
+			} else { //Pre-2.6 support when mtype is not passed
```
[the remark this answers is not in the document]
Comment (vrtcal-dev) [reply adapters/vrtcal/vrtcal.go]: Cleaned up this function to reflect your suggestions and updated related test cases.

**Review by onkarvhanumante — APPROVED** [review]

**Review by gargcreation1992 — APPROVED** [review]

**Pull request merged**